src/runtime/libuv.cpp

```diff
@@ -2,21 +2,231 @@
 Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 
-Author: Markus Himmel
-*/
+Author: Markus Himmel, Sofia Rodrigues
+ */
+#include <pthread.h>
+
+#include "runtime/uv_scheduler.h"
 #include "runtime/libuv.h"
+#include "runtime/object.h"
+
+namespace lean {
 
 #ifndef LEAN_EMSCRIPTEN
 #include <uv.h>
 
+scheduler_t global_scheduler;
+
+static void lean_uv_timer_finalizer(void* ptr) {
+    lean_uv_timer_object* timer_obj = (lean_uv_timer_object*) ptr;
+    timer_obj->m_uv_timer.data = ptr;
+
+    scheduler_lock(&global_scheduler);
+
+    uv_close((uv_handle_t*) & timer_obj->m_uv_timer, [](uv_handle_t* handle) {
+        free((lean_uv_timer_object*)handle->data);
+    });
+
+    if (timer_obj->m_promise != NULL) lean_dec(timer_obj->m_promise);
+
+    scheduler_unlock(&global_scheduler);
+}
+
+extern "C" void initialize_libuv() {
+    g_uv_timer_external_class = lean_register_external_class(lean_uv_timer_finalizer, noop_foreach);
+    scheduler_init(&global_scheduler);
+    lthread([]() { scheduler_run_loop(&global_scheduler); });
+}
+
 extern "C" LEAN_EXPORT lean_obj_res lean_libuv_version(lean_obj_arg o) {
     return lean_unsigned_to_nat(uv_version());
 }
 
+/* UV.Loop.configure (options : Loop.Options) : BaseIO Unit */
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_event_loop_configure(b_obj_arg options, obj_arg /* w */ ) {
+    bool accum = lean_ctor_get_uint8(options, 0);
+    bool block = lean_ctor_get_uint8(options, 1);
+
+    scheduler_lock(&global_scheduler);
+
+    if (accum && uv_loop_configure(global_scheduler.loop, UV_METRICS_IDLE_TIME) != 0) {
+        return io_result_mk_error("failed to configure global_scheduler.loop with UV_METRICS_IDLE_TIME");
+    }
+
+    #if!defined(WIN32) && !defined(_WIN32)
+    if (block && uv_loop_configure(global_scheduler.loop, UV_LOOP_BLOCK_SIGNAL, SIGPROF) != 0) {
+        return io_result_mk_error("failed to configure global_scheduler.loop with UV_LOOP_BLOCK_SIGNAL");
+    }
+    #endif
+
+    scheduler_unlock(&global_scheduler);
+
+    return lean_io_result_mk_ok(lean_box(0));
+}
+
+/* UV.Loop.alive : BaseIO UInt64 */
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_event_loop_alive(obj_arg /* w */ ) {
+    scheduler_lock(&global_scheduler);
+    int is_alive = uv_loop_alive(global_scheduler.loop);
+    scheduler_unlock(&global_scheduler);
+
+    return lean_io_result_mk_ok(lean_box(is_alive));
+}
+
+void handle_timer_event(uv_timer_t* handle) {
```

Comment:
The `running` variable checks if we solved the current Promise and `start` if the uv timer is running, so I think that I can remove the start variable in some sense but it would cause the `stop` function to not run the next callback after the stop and the promise that it is holding will never be resolved.

Reply:
You can check if the promise is resolved using `lean_io_get_task_state_core`.